python.rst

```diff
@@ -2,36 +2,96 @@
 ====================
 
-Goal
+Background
 ====
 
-The goal of this style-guide is to standardize the motif of Python at
-Oscar in a way that reduces the risk of bugs while increasing our
-ability to utilize the richer features of Python when necessary. In
-addition, it reduces the risk exposure from the many third party
-libraries we use. Finally, if it is true that reading code is 90% of
-programming, then we should favor readability over writability.
+The goal of this document is to guide the style of Python code at
+Oscar to reduce bugs and improve readability.
 
-Example: `Import packages and modules only, do not import objects or
-functions directly`_ may seem constrictive or verbose, but it allows
-objects in module namespaces to be mutated during execution without
-unexpected behavior due to object duplication. Examples of this in our
-current code-base are sqlalchemy-continuum and tagging.
+Many teams use the `black <https://github.com/psf/black>`_
+auto-formatter to avoid having to worry (or even worse, argue!) about
+formatting. We recommend using it, but it does not cover everything
+here.
 
-Code of Conduct
+Code of conduct
 ===============
 
 - Be polite and respectful.
 - Don’t make assumptions—be clear and concise.
-- Your co workers deserve a response to comments, questions or
-  concerns, even if it is to respectfully decline or dissent.
+- Your coworkers deserve a response to comments, questions or concerns, even if it is to respectfully decline or dissent.
 - Boy Scout Rule: Always check a module in cleaner than you checked it
   out. Clean-up and compliance to the style guide is the
-  responsibility of whomever touches the code.
+  responsibility of whoever touches the code.
+- Don't Boil the Ocean: if you weren't already working on a module for a diff, you
+  probably shouldn't be cleaning it up. It confuses reviewers and makes it harder
+  to spot bugs. (Feel free to open a separate diff for the clean-up, though!)
+- Don't be afraid to ask for help or clarification. We're all
+  learning.
 
 Language Rules
 ==============
 
+Testing
+-------
+
+Use pytest-style asserts in unit tests
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+- Use ``assert`` statements in unit tests. `Our test runner
+  <https://docs.pytest.org/en/latest/assert.html>`_ will intelligently
+  report the failure, so you can avoid the many names of JUnit
+  legacy methods (self.assert*). It's easier to read and write.
+
+    .. code-block:: python
+
+        # Good
+        assert foo == bar
+        assert foo != bar
+        assert foo in bar
+        assert {1, 2, 3} == {3, 2, 1}
+        with pytest.raises(SomeException):
+            do_something()
+
+        # Bad
+        self.assertEqual(foo, bar)
+        self.assertNotEqual(foo, bar)
+        self.assertIn(foo, bar)
+        self.assertSetEqual({1, 2, 3}, {3, 2, 1})
+        with self.assertRaises(SomeException):
+            do_something()
+
+- Parameterize tests instead of copy-pasting or using
```

Comment:
For loops? fine. But I prefer DAMP to DRY unit tests. Using parameters in this way obfuscates what any of those values represent.

Reply:
You're right. Pytest's built-in parameterization is better in this regard:

```
@pytest.mark.parameterize(["base","exponent","expected"], [(2, 2, 4), (2, 3, 8), (1,9,1), …])
def test_pow(base, exponent, expected):
    blah
```

—but that doesn't work with unittest.TestCase-style tests, which almost all of ours are. There's another API for parameterized, which I don't love the noise level of:

```
from parameterized import parameterized, param

@parameterized([
    param(base=2, exponent=2, expected=4),
    param(base=2, exponent=3, expected=8),
    …
])
def test_pow(base, exponent, expected):
    blah
```

—but the ability to fail those cases independently probably makes it still worth it over for-loops.

Thoughts on the param version?

Reply:
I like it! I still don't think the copy/pasta versions are strictly negative (especially if people end up jumping through hoops or have weird control logic based on the parameters), but if you're going to use parameterized, this is definitely nice.

Reply:
ok, I'll update the PR accordingly.

And I agree with the "copy/pasta [when you] have weird control logic based on the parameters"; I'll add some note to that effect. This was top of mind for me because Pandas test code suffers more from excessive (i.e. same control logic) copy/pasta drowning the differences in noise than it does from cleverness obscuring the signal.